File: 0002_codes/C - Kernel_Level/OS02.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
 * OS02 : path breakdown tool
 *
 * This program:
 * - reads the PATH environment variable
 * - splits it into separate directories
 * - prints each directory on a new line
 */

int main(void) {
    const char *path = getenv("PATH");

    printf("=========================================\n");
    printf("              OS02 - path                \n");
    printf("=========================================\n");

    if (path == NULL) {
        printf("PATH variable not found\n");
        printf("=========================================\n");
        return 1;
    }

    /* make a copy of PATH because strtok modifies the string */
    char *path_copy = malloc(strlen(path) + 1);
    if (path_copy == NULL) {
        printf("memory allocation failed\n");
        return 1;
    }
    strcpy(path_copy, path);

    printf("directories in PATH:\n\n");

    /* split using ':' on linux/mac, ';' on windows */
    const char *delimiter;

#if defined(_WIN32) || defined(_WIN64)
    delimiter = ";";
#else
    delimiter = ":";
#endif

    char *token = strtok(path_copy, delimiter);

    while (token != NULL) {
        printf("%s\n", token);
        token = strtok(NULL, delimiter);
    }

    free(path_copy);

    printf("\n=========================================\n");

    return 0;
}

File: 0002_codes/JAVA - User_Level/OS02.java
import java.util.Scanner;

/*
 * OS02 - path breakdown tool
 * this program:
 * - reads the PATH environment variable
 * - splits it into separate directories
 * - prints each directory on a new line
 */

public class OS02 {

    public static void main(String[] args) {

        System.out.println("=========================================");
        System.out.println("              OS02 - path                ");
        System.out.println("=========================================");

        /* get PATH variable */
        String path = System.getenv("PATH");

        if (path == null) {
            System.out.println("PATH variable not found");
            System.out.println("=========================================");
            return;
        }

        /* determine path separator */
        String separator;

        String osName = System.getProperty("os.name", "").toLowerCase();
        if (osName.contains("win")) {
            separator = ";";
        } else {
            separator = ":";
        }

        System.out.println("directories in PATH:\n");

        /* split and print each directory */
        String[] dirs = path.split(separator);

        for (String dir : dirs) {
            System.out.println(dir);
        }

        System.out.println("\n=========================================");
    }
}
